### ISSUE TITLE
The is no (configurable) delay while typing.

### ISSUE THREAD

**Issue by strpbrk** [clangd/vscode-clangd#262]
Allow me to ask for this feature, even though I fear that this feature would need to be implemented in the language client project. Nevertheless the configuration option would need to be supported here...

There doesn't seem to be a configurable, or at least **significant** delay while typing. I would expect that semantic information, or at the very least the list of found **problems** would not jump around while I'm typing. If the former is necessary, at least the latter should not be. The subjective feeling is that I'm working on an old computer with unnecessary distractions jumping all over the screen. (I have auto suggestions and all auto stuff turned off).

If there is nothing you can do, could you please point me to other resources, such as open issues on the language client project, a missing feature elsewhere, or at least an explanation why this would not be possible any time soon. I have tried, but my search has not found anything of significance.

Thanks!

**Comment by sam-mccall** [comment]
The server is in charge of publishing diagnostics. The client could suppress updates to diagnostics while typing - this seems likely to be a vscode-level feature rather than even the vscode-languageclient library. I don't have any particular insight there.

But the server can also delay computing/pushing diagnostics while it's seeing updates from the client.
We actually do have this code in clangd (a debounce delay before reparsing for diagnostics on the server side), but it's tuned for responsiveness: the delay is equal to the parse time for the file, but never less than 50 or more than 500ms.

Before exposing this it would probably be nice to know whether it actually does what you want: these things are about "feel" and hard to predict. 
If you're able to build clangd locally, you could add this line to `clang-tools-extra/clangd/tool/ClangdMain.cpp`, near where other options are set:
```
Opts.UpdateDebounce = DebouncePolicy::fixed(std::chrono::seconds(5));
```

**Comment by strpbrk** [comment]
I have built and tested clangd. Verified that the option is correctly used in "TUScheduler.cpp" to create a Deadline.
This had no effect on the output, certainly no delay of 5s.

However, even if this worked, I would have to disagree that it is a correct solution.
The server should not be concerned about things that alter the user experience. It simply provides a service, and it is just fine that it is tuned for responsiveness.
It is up to the client to do one of two things (or both):
1. Send text changes after a short delay, or after the user stops typing.
2. Apply updates from the server with a delay, or after the user stops typing.
Microsoft's plugin has "intelliSenseUpdateDelay" option to control one of these (although I didn't try their plugin).

I have poked around the code, and have found that the first feature could potentially be implemented by providing a "middleware.didChange" handler that would delay before invoking "next". I got a simple delay working as a proof-of-concept, and it indeed holds before sending changes to the server. The code in the language client library seems dirty and not well suited for a proper implementation that should group changes and flush them on a completion request, but newer versions seem to have improved this.

I have to say that I find it very weird that such a great plugin would generate and send one event per key press to the language server (unless you are cutting or pasting selections), although it does indeed seem it is an issue for the language client library.

**Comment by strpbrk** [comment]
I have implemented this feature, and it seems to be working properly, at least for my environment.
It uses latest dependencies, so I cannot create a pull request.

Issue #151 is similar to this one, but it only complains about flashing status bar, not for the reason behind it. The status bar will be silent with this change... as it should.

Please find changes in my forked repository: https://github.com/strpbrk/vscode-clangd

**Comment by HighCommander4** [comment]
> It uses latest dependencies, so I cannot create a pull request.

We do update dependencies from time to time, so you could create a pull request that also updates dependencies.

**Comment by strpbrk** [comment]
Let me use the plugin for a while, and I will do that. I just made a significant simplification related to my initial version anyway.
In the meantime, it is there for anyone to try...

**Comment by edding4500** [comment]
Hey guys, I just ran into the same issue: on a rather old notebook the constant LSP calls cause quite some heavy load.
I did a similar implementation to that of @strpbrk but ditched it for his version as that one looked cleaner.
On the current main branch @strpbrk's changes don't need dependency updates anymore. I did however notice some jumping inline decorations and breaking syntax highlighting. @strpbrk can you confirm this?

**Comment by strpbrk** [comment]
Thank you for the kind words, @edding4500. So far, I have been using plugin v0.1.17 with my modifications.

There is an issue with my changes - if an edit is made through an operation, such as refactoring or using a code fix, while there are pending edits in affected lines, it might mess up the code in those lines. Simply editing a file, including copy/paste of blocks of text, works great for me. Before refactoring, I just save the file, which flushes pending changes.

To fix this, the same thing "didSave" does should be done before an operation is performed, but I didn't find an appropriate extension point, and piercing a few frameworks would just make this pull request even less desirable. But, I don't think this is what you are describing...

When I find a bit of time, I will sync up to the master and see what happens, and will also look at changes in dependencies. Feel free to describe your issues in more detail, or maybe make a video or a gif. What I have been using has no distracting features, as long as I save before refactoring.

**Comment by pierzchalski** [comment]
@strpbrk did you end up finding time to sync up with master? I too would like to stop waiting for `clangd` to error out on `incomplete type v`, `incomplete type vo`, `incomplete type voi`...

**Comment by strpbrk** [comment]
@pierzchalski Sorry. I took a look a few months ago and gave up on this. With the current (at that time) API, even if it would be possible, the code would be unmaintainable. Unless there are other hooks that I missed.

It is definitely a client problem and should not be patched up by changes in clangd server.

**Comment by HighCommander4** [comment]
> [@pierzchalski](https://github.com/pierzchalski) Sorry. I took a look a few months ago and gave up on this. With the current (at that time) API, even if it would be possible, the code would be unmaintainable. Unless there are other hooks that I missed.
> 
> It is definitely a client problem and should not be patched up by changes in clangd server.

Perhaps you could file an issue in an appropriate upstream repository (either https://github.com/microsoft/vscode-languageserver-node/ or https://github.com/microsoft/vscode, you may have a better idea based on explorations which one)?

**Comment by Iritscen** [comment]
> There is an issue with my changes - if an edit is made through an operation, such as refactoring or using a code fix, while there are pending edits in affected lines, it might mess up the code in those lines. Simply editing a file, including copy/paste of blocks of text, works great for me. Before refactoring, I just save the file, which flushes pending changes.

I'm just wading into this, but looking at the code on VS Code's side, what about this option for sending the full document instead of incremental changes to the server? https://github.com/microsoft/vscode-languageserver-node/blob/a7c5e6557209ef416526fe509296ee00179c9aa3/protocol/src/common/protocol.ts#L1894

I'm not clear on whether this is a choice that can be made by the plugin, but if the clangd plugin can ask for Full document sync then there is no need to accumulate changes and potentially miss something when an operation is performed. Can that be done through here?: https://github.com/microsoft/vscode-languageserver-node/blob/a7c5e6557209ef416526fe509296ee00179c9aa3/protocol/src/common/protocol.ts#L2061

**Comment by Iritscen** [comment]
I might be!  But I have no experience developing VSC plugins, so I was wondering if strpbrk wanted to take one more try at it first.

**Comment by Iritscen** [comment]
All right, I've been working on this, and I have a solution that seems pretty solid and which isn't dependent on catching all change events before they leave the editor.  Instead of intercepting changes on their way out to clangd, it intercepts the diagnostics on the way back in and simply doesn't show them unless the user-configured time has passed.  I was seeing a strange issue with broken syntax coloring when using strpbrk's approach, which I think was caused by clangd not having parsed the latest code and returned the results to VSC.  Because clangd still processes the code in realtime with my approach, this doesn't happen.

I also added a second option besides a fixed delay which might be more desirable for some coders, which is to wait until they leave the current line before showing diagnostics for it.  Sometimes you simply want to sit on a line and think about what to write next without errors appearing; also, if you're waiting on a completion suggestion then it's nice to not get an error on your half-typed text in the meantime.

To be clear, I didn't end up using the Full document sync method because (a) the sync method has to be set on clangd's side, not the plugin's, and (b) it doesn't help in implementing my above solution anyway.  So only the changes to the active document are being sent to clangd, as usual, not the whole document.

This does mean that clangd is still processing the source file every single time a character is typed, which some have objected to on an efficiency level, but when I looked into the possibility of adding this delay on clangd's side by collecting incoming change events or outgoing work products in a queue, it was considerably more complicated due to the threading and queuing which clangd already employs.  A lot of changes would need to be made to the existing logic – even when utilizing the existing debounce delay feature – and I felt that I would likely cause regressions.  This approach seems a lot safer.  I've tested it a bit and wasn't able to break anything, but please let me know if there are concerns about my method or if anyone finds a bug.  I'm very much a newbie to this whole environment so I might have made some novice errors.

https://github.com/Iritscen/vscode-clangd/tree/iritscen/delayed-diagnostics

**Comment by PeterFeicht** [comment]
> Could you file a new issue about this please? I think this should be fixable on the server side, thereby mitigating the responsiveness you're experiencing to potentially a significant extent.

Sure. Where should I put it, this repo or somewhere else?

**Comment by Iritscen** [comment]
> I guess this could technically lead to a slight responsiveness regression for the code completion request in the case where you're triggering it explicitly with Ctrl+Enter or such, since clangd is only going to start parsing the file after the queue has been flushed in response to the Ctrl+Enter (whereas before, clangd would have started parsing after your last keystroke that actually _modified_ the file).

I feel that the extension might benefit from employing both approaches – holding onto outgoing changes _and_ holding onto incoming diagnostics.  For implementing a fixed-time diagnostic delay, only the outgoing queue need be used, and incoming diagnostics would be shown immediately.  But if you do choose to support the alternative options of line-change and file-save diagnostic delay, I think you'd want to employ my approach of holding the incoming diagnostic updates in a separate bucket and allowing through the completion suggestions immediately.  That way, if the user has on-line-change or on-file-save diagnostic delay enabled, they won't need to manually trigger the suggestions in order to see them and won't need to worry about seeing diagnostics update too when they do that.  They will only see diagnostics update upon their chosen event, and meanwhile completion suggestions will only be delayed by whatever the debounce is set to (and they can always lower that setting all the way to 0ms if they desire).
